### DISCUSSION TITLE
Help needed: Fixing performance problems w/js_sys::Math::random

### DISCUSSION THREAD

**Discussion by NateLevin1** [wasm-bindgen/wasm-bindgen#3196] [Q&A] [answered]
Hey everybody, I'm looking for some help fixing a performance problem in some Rust WASM code I am writing.

I'm building a simulation program that needs to calculate **a ton** of random numbers — I'm talking in the range of 16 to 144 million or more. I've benchmarked my code with the Chrome profiler, and it says that calls to `js_sys::Math::random` are taking up ~1.3s of blocking CPU time. I'm pretty certain I know why this is happening, the WASM is constantly having to call the JS glue code to run Math.random which slows everything down.

Is there a way I can re-architect my code to have better performance? Is there an alternate random number generator that will be good for my use case?

Any help would be greatly appreciated 😄

**Comment by Liamolucko** [comment, marked as answer]
You could give the [`rand`](https://docs.rs/rand) crate a go, which includes random number generators written in Rust that could potentially be faster than using `Math.random`. To seed them from the browser's RNG, you'll need to enable the `js` feature of the `getrandom` crate, like so:

```toml
# Cargo.toml
[dependencies]
rand = "0.8.5"
getrandom = { version = "0.2.8", features = ["js"] }
```

`Math.random` wouldn't be slow because of the JS glue code, though, because it doesn't have any - all it does is return a single `f64`, which wasm can accept directly without any glue.

I haven't actually checked whether `rand` is faster. If you don't need a cryptographic RNG, [`rand::rngs::SmallRng`](https://docs.rs/rand/latest/rand/rngs/struct.SmallRng.html) should also be faster than its default RNG.

**Reply by NateLevin1** [reply]
Thanks for the reply — this seems super useful, though I’m having one slightly unrelated problem. I need a random f64 between 0-1, but SmallRng seems to only be able to give random u32/u64 values. Is there an idiomatic way to convert between these, or should I use a different generator?

Thanks again for the help!

**Reply by Liamolucko** [reply]
Yes, there's an idiomatic way to convert - you need to pick a [`Distribution`](https://docs.rs/rand/latest/rand/distributions/trait.Distribution.html) to use, which defines how to make random values from raw random bytes. In your case, you need the [`Standard`](https://docs.rs/rand/latest/rand/distributions/struct.Standard.html) distribution, which samples numbers from 0-1 the same as `Math.random`. To actually use it, you write `rng.sample(Standard)`, where `rng` is the RNG you've already constructed. You could also use `rng.gen()`, which is shorthand for the same thing.
